src/core/frames/frame_controller.ts

```diff
@@ -288,16 +301,25 @@ export class FrameController
           const response = { statusCode, redirected, responseHTML }
 
           session.visit(frame.src, {
-            action,
+            action: this.action,
             response,
             visitCachedSnapshot,
             willRender: false,
+            updateHistory: false,
+            restorationIdentifier: this.restorationIdentifier,
           })
         }
       }
     }
   }
 
+  changeHistory() {
+    if (this.action && this.frame) {
+      const method = getHistoryMethodForAction(this.action)
+      session.history.update(method, expandURL(this.frame.src as string), this.restorationIdentifier)
```

Comment:
What makes the `as string` casting necessary?

Is it possible that `this.frame.src` is `null`? Could `|| ""` cover that case without circumventing the compiler?